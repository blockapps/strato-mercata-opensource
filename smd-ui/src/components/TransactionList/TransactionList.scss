.tx-row {
  margin: 2%; }
