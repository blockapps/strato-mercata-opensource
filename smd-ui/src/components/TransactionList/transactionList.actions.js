export const FETCH_TX = 'FETCH_TX';
export const FETCH_TX_SUCCESSFUL = 'FETCH_TX_SUCCESSFUL';
export const FETCH_TX_FAILED = 'FETCH_TX_FAILED';
export const UPDATE_TX = 'UPDATE_TX';
export const PRELOAD_TX = 'PRELOAD_TX';

export const updateTx = function (data) {
  return {
    type: UPDATE_TX,
    data
  }
}

export const preloadTx = function (data) {
  return {
    type: PRELOAD_TX,
    data
  }
}

export const fetchTx = function (last, chainId) {
  return {
    type: FETCH_TX,
    last: last,
    chainId: chainId
  }
};

export const fetchTxSuccess = function (res) {
  return {
    type: FETCH_TX_SUCCESSFUL,
    tx: res
  }
};

export const fetchTxFailure = function (error) {
  return {
    type: FETCH_TX_FAILED,
    error: error,
  }
};
