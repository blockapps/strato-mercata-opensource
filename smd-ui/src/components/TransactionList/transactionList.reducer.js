import {
  FETCH_TX,
  FETCH_TX_SUCCESSFUL,
  FETCH_TX_FAILED,
  UPDATE_TX,
  PRELOAD_TX
} from './transactionList.actions';

const initialState = {
  tx: [],
  transactions: [],
  error: null,
};

const reducer = function (state = initialState, action) {
  switch (action.type) {
    case UPDATE_TX:
      return {
        ...state,
        transactions: action.data,
      }
    case PRELOAD_TX:
      return {
        ...state,
        transactions: action.data,
      }
    case FETCH_TX:
      return {
        ...state,
        tx: state.tx,
        error: null,
      };
    case FETCH_TX_SUCCESSFUL:
      return {
        ...state,
        tx: action.tx,
        error: null,
      };
    case FETCH_TX_FAILED:
      return {
        ...state,
        tx: state.tx,
        error: action.error
      };
    default:
      return state;
  }
};

export default reducer;
