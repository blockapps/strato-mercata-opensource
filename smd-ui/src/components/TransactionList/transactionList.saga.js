import {
  takeEvery,
  put,
  call
} from 'redux-saga/effects';
import {
  FETCH_TX,
  fetchTxSuccess,
  fetchTxFailure
} from './transactionList.actions';
import { env } from '../../env';
import { handleErrors } from '../../lib/handleErrors';
import { createUrl } from '../../lib/url';

export function getTx(last, chainid) {
  const options = { query: { chainid } };
  const url = env.STRATO_URL + createUrl("/transaction/last/15", options);

  return fetch(
    url,
    {
      method: 'GET',
      credentials: "include",
      headers: {
        'Accept': 'application/json'
      },
    })
    .then(handleErrors)
    .then(function (res) {
      return res.json();
    })
    .catch(function (error) {
      throw error;
    });
}

export function* fetchTx(action) {
  try {
    let response = yield call(getTx, action.last, action.chainId);
    yield put(fetchTxSuccess(response));
  }
  catch (err) {
    yield put(fetchTxFailure(err));
  }
}

export default function* watchFetchTx() {
  yield takeEvery(FETCH_TX, fetchTx);
}
