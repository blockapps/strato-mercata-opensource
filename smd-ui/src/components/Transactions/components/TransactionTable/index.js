import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Field, reduxForm, reset } from 'redux-form';
import { TRANSACTION_QUERY_TYPES, RESOURCE_TYPES } from '../../../QueryEngine/queryTypes';
import { updateQuery, clearQuery, executeQuery, removeQuery} from '../../../QueryEngine/queryEngine.actions';
import { withRouter } from 'react-router-dom';
import { Text, Position, Tooltip, Button } from '@blueprintjs/core';
import { parseDateFromString } from '../../../../lib/dateUtils';
import mixpanelWrapper from '../../../../lib/mixpanelWrapper';
import { fetchTx } from '../../../TransactionList/transactionList.actions';
import HexText from '../../../HexText';

class TransactionTable extends Component {
  componentDidMount() {
    this.props.fetchTx(null, this.props.selectedChain);
    this.props.executeQuery(RESOURCE_TYPES.transaction, this.props.query, this.props.selectedChain);
  }

  componentWillReceiveProps(newProps) {
    if (newProps.query !== this.props.query) {
      newProps.executeQuery(RESOURCE_TYPES.transaction, newProps.query, newProps.selectedChain);
    }
    if (newProps.selectedChain !== this.props.selectedChain) {
      this.props.fetchTx(null, newProps.selectedChain);
      newProps.executeQuery(RESOURCE_TYPES.transaction, newProps.query, newProps.selectedChain);
    }
  }

  updateQuery = (values) => {
    if (values.query && values.value) {
      this.props.updateQuery(values.query, values.value);
      this.props.dispatch(reset('transaction-query'));
    }
  }

  // dispatchSubmit = () => {
  //   this.props.dispatch(submit('transaction-query'));
  // }

  refresh = () => {
    this.props.clearQuery();
    this.props.executeQuery(RESOURCE_TYPES.transaction, this.props.query, this.props.selectedChain);
  };
  
  render() {
    const { handleSubmit, history} = this.props; 
    const handleClick = function(hash) {
      mixpanelWrapper.track('transactions_row_click');
      history.push(`/transactions/${hash}`);
    }
    let txRows = this.props.queryResults.length && this.props.queryResults[0]['transactionType'] && this.props.queryResults.map(
      function (tx, i) {
        return (
          <tr key={i} onClick={() => {
            handleClick(tx.hash)
          }}>
            <td width="40%">
              <HexText value={tx.hash} classes="small smd-pad-4" />
            </td>
            <td width="10%">
              <Text ellipsize={true}>
                <Tooltip tooltipClassName="smd-padding-8" content={tx.value + ' wei'} position={Position.TOP_LEFT}>
                  <small>{tx.value} wei</small>
                </Tooltip>
              </Text>
            </td>
            <td width="10%">
              <Text ellipsize={true}>
                <small>
                  {tx.blockNumber}
                </small>
              </Text>
            </td>
            <td width="20%">
              <Text ellipsize={true}>
                <small>
                  {parseDateFromString(tx.timestamp)}
                </small>
              </Text>
            </td>
            <td width="20%">
              <small>{tx.transactionType}</small>
            </td>
          </tr>
        )
      }
    );

    const required = value => value ? undefined : 'Required'
    const queryTypes = TRANSACTION_QUERY_TYPES;
    const queryForm =
      <div className="row smd-pad-4">
        <div className="col-sm-12">
          <form onSubmit={handleSubmit(this.updateQuery)}>
            <div className="pt-control-group smd-full-width">
              <div className="pt-select" id="tour-query-type">
                <Field
                  type="select"
                  component="select"
                  placeholder="Query Type"
                  name="query"
                  validate={required}
                  required
                >
                  {
                    Object.getOwnPropertyNames(queryTypes).map(function (name) {
                      return <option key={name} value={queryTypes[name].key}>{queryTypes[name].displayName}</option>
                    })
                  }
                </Field>
              </div>
              <div className="smd-input-width">
                <Field
                  type="text"
                  className="pt-input pt-fill"
                  component="input"
                  name="value"
                  placeholder="Query Term"
                  validate={required}
                  onKeyPress={
                    (e) => {
                      if (e.key === 'Enter') {
                        //this.dispatchSubmit();
                        mixpanelWrapper.track('transactions_query_submit');
                      }
                    }
                  }
                  dir="auto" />
              </div>
              <Button type="submit" onClick={() => {
                //this.dispatchSubmit();
                mixpanelWrapper.track('transactions_query_submit');
              }}
                className="pt-intent-primary pt-icon-arrow-right" />
            </div>
          </form>
        </div>
      </div>

    const query = this.props.query;
    const removeQuery = this.props.removeQuery;
    const tags = Object.getOwnPropertyNames(query).map((queryType, i) => {
      const queryValue = query[queryType];
      return (
        <span key={'tag-' + queryType + '-' + i} className="pt-tag pt-tag-removable smd-margin-right-4">
          {queryType + ': ' + queryValue}
          <button onClick={() => {
            removeQuery(queryType);
            mixpanelWrapper.track('transactions_query_remove_tag');
          }} className="pt-tag-remove" />
        </span>
      )
    });

    const queries =
      <div>
        {queryForm}
        <div className="row smd-pad-4">
          <div className="col-sm-12">
            {tags}
          </div>
        </div>
      </div>

    return (
      <div className="pt-card pt-dark pt-elevation-2">
        <div className="row smd-pad-4">
          <div className="col-sm-11 text-left">
            <span className="h3">Query Builder</span>
          </div>
          <div className="col-sm-1 text-right">
            <Button onClick={this.refresh} className="pt-intent-primary pt-icon-refresh" />
          </div>
        </div>

        {queries}

        <div className="row">
          <div className="col-sm-12">
            <table className="pt-table pt-interactive pt-condensed pt-striped"
              style={{ tableLayout: 'fixed', width: "100%" }}>
              <thead>
                <tr>
                  <th width="40%"><h5>Hash</h5></th>
                  <th width="10%" className="text-right"><h5>Value</h5></th>
                  <th width="10%"><h5>Block Number</h5></th>
                  <th width="20%"><h5>Timestamp</h5></th>
                  <th width="20%"><h5>Type</h5></th>
                </tr>
              </thead>

              <tbody>
                {!txRows ? <tr>
                  <td colSpan={5}>No Data</td>
                </tr> : txRows}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    );

  }
}

export function mapStateToProps(state) {
  return {
    query: state.queryEngine.query,
    queryResults: state.queryEngine.queryResult,
    selectedChain: state.chains.selectedChain
  };
}
const formed = reduxForm({ form: 'transaction-query' })(TransactionTable);
const connected = connect(mapStateToProps, {
  updateQuery,
  removeQuery,
  executeQuery,
  clearQuery,
  fetchTx
})(formed);
export default withRouter(connected);
