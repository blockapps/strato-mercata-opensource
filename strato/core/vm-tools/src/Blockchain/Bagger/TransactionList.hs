module Blockchain.Bagger.TransactionList
  ( TransactionList,
    -- , emptyTransactionList
    singletonTransactionList,
    insertTransaction,
    trimBelowNonce,
    trimAboveCost,
    popSequential,
    toList,
  )
where

import Blockchain.Data.TransactionDef
import Blockchain.Sequencer.Event (OutputTx (..))
import Data.Foldable (foldl')
import qualified Data.Map.Strict as M

type TransactionList = M.Map Integer OutputTx

nonce :: OutputTx -> Integer
nonce = transactionNonce . otBaseTx

gasPrice :: OutputTx -> Integer
gasPrice = transactionGasPrice . otBaseTx

--emptyTransactionList :: TransactionList
--emptyTransactionList = M.empty

singletonTransactionList :: OutputTx -> TransactionList
singletonTransactionList t = M.singleton (transactionNonce $ otBaseTx t) t

-- should replace TXs with identical nonces but different gas cost to one with higher gas cost
-- returns (Maybe <txThatWasReplaced/txToDropFromSeen>, newTransactionList)
insertTransaction :: OutputTx -> TransactionList -> (Maybe OutputTx, OutputTx, TransactionList)
insertTransaction t tl =
  let nonce' = nonce t
      oldTx = M.lookup nonce' tl
   in case oldTx of
        Nothing -> (Nothing, t, M.insert nonce' t tl)
        Just existing ->
          if gasPrice existing <= gasPrice t
            then (oldTx, t, M.insert nonce' t tl)
            else (Just t, existing, tl)

trimBelowNonce :: Integer -> TransactionList -> ([OutputTx], TransactionList)
trimBelowNonce nonce' tl = let (lt, gte) = M.partitionWithKey (\k _ -> k < nonce') tl in (M.elems lt, gte)

trimAboveCost :: Integer -> (OutputTx -> Integer) -> TransactionList -> ([OutputTx], TransactionList)
trimAboveCost maxCost calcCost tl =
  let (tooHigh, justRight) = M.partitionWithKey (\_ v -> calcCost v > maxCost) tl in (M.elems tooHigh, justRight)

popSequential :: Integer -> TransactionList -> ([OutputTx], TransactionList)
popSequential nonce' tl = (popped, M.fromList kept)
  where
    (_, popped, kept) = foldl' theFold initialFoldState (M.toAscList tl)
    initialFoldState = (nonce' - 1, [], [])
    theFold (lastNonce, popped', kept') e@(elemNonce, elemTx) =
      if elemNonce == lastNonce + 1
        then (elemNonce, elemTx : popped', kept')
        else (lastNonce, popped', e : kept')

toList :: TransactionList -> [OutputTx]
toList = M.elems
