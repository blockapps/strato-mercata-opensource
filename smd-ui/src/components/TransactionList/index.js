import React, {Component} from 'react';
import {connect} from 'react-redux';
import {withRouter} from 'react-router-dom';
import './TransactionList.css'
import { Text } from '@blueprintjs/core';
import mixpanelWrapper from '../../lib/mixpanelWrapper';
import { parseDateFromString } from '../../lib/dateUtils';
import HexText from '../HexText';
import { GET_TRANSACTIONS } from '../../sockets/rooms';
import { subscribeRoom, unSubscribeRoom } from '../../sockets/socket.actions'

class TransactionList extends Component {

  componentDidMount() {
    this.props.subscribeRoom(GET_TRANSACTIONS)
  }

  componentWillUnmount() {
    this.props.unSubscribeRoom(GET_TRANSACTIONS)
  }

  render() {
    const self = this;
    let txRows = this.props.transactions !== undefined && this.props.transactions.slice(0, 5).map(
      function (tx, i) {        
        return (
          <tr
            key={i}
            onClick={e => {mixpanelWrapper.track("dashboard_transaction_click"); self.props.history.push('/transactions/' + tx.hash)}}
          >
            <td width="40%">
              <HexText value={tx.hash} classes="small smd-pad-4"/>
            </td>
            <td width="23%" className="text-right">
              <small>{tx.value}</small>
            </td>
            <td width="22%">
              <Text ellipsize={true}>
                <small>
                  {parseDateFromString(tx.timestamp)}
                </small>
              </Text>
            </td>
            <td width="15%">
              <small>{tx.transactionType}</small>
            </td>
          </tr>
        )
      }
    );

    return (
      <div className="pt-card pt-dark pt-elevation-2">

        <h4>Recent Transactions (Main Chain)</h4>
        <table className="pt-table pt-interactive pt-condensed pt-striped" style={{tableLayout: 'fixed', width: '100%'}}>
          <thead>
          <tr>
            <th width="40%"><h5>Hash</h5></th>
            <th className="text-right" width="23%"><h5>Value</h5></th>
            <th width="22%"><h5>Timestamp</h5></th>
            <th width="15%"><h5>Type</h5></th>
          </tr>
          </thead>

          <tbody>
          {txRows.length === 0 ? <tr><td colSpan={5}>No Data</td></tr> : txRows}
          </tbody>
        </table>
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    transactions: state.transactions.transactions
  };
}

export default withRouter(connect(mapStateToProps, {subscribeRoom,unSubscribeRoom})(TransactionList));
