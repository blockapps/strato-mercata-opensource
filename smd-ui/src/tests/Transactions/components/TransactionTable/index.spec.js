import React from 'react'
import TransactionTable, { mapStateToProps } from '../../../../components/Transactions/components/TransactionTable/index'
import { mount } from 'enzyme'
import { reducer as formReducer } from 'redux-form'
import { createStore, combineReducers } from 'redux'
import { Provider } from 'react-redux'
import { transactions } from '../../../TransactionList/transactionListMock'

describe("TransactionTable: index", () => {

  let store

  beforeEach(() => {
    store = createStore(combineReducers({ form: formReducer }))
  })

  test('simulate buttons click', () => {
    const props = {
      query: { last: 15 },
      queryResults: transactions,
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      history: { push: jest.fn() }
    }
    let wrapper = mount(
      <Provider store={store}>
        <TransactionTable.WrappedComponent {...props} />
      </Provider>
    )
    wrapper.find('button').at(0).simulate('click');
    expect(props.clearQuery).toHaveBeenCalled();

    wrapper.find('button').at(2).simulate('click');
    expect(props.removeQuery).toHaveBeenCalled();

    wrapper.find('tr').at(2).simulate('click')
    expect(props.history.push).toHaveBeenCalled()

  });

  test('simulate form submit', () => {
    const props = {
      query: {},
      queryResults: transactions,
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      handleSubmit: jest.fn(),
      history: { push: jest.fn() }
    }
    let wrapper = mount(
      <Provider store={store}>
        <TransactionTable.WrappedComponent {...props} />
      </Provider>
    )
    const t = wrapper.find('Field').last().simulate('keyPress', { target: { key: 'Enter', value: 'temp' } });
    const test = wrapper.find('Button').last().simulate('click');
    const r = wrapper.find('Form').simulate('submit');
    expect(props.handleSubmit).toHaveBeenCalled();

  });

  test('component methods', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn().mockReturnValue('executeQuery'),
      updateQuery: jest.fn().mockReturnValue('updateQuery'),
      removeQuery: jest.fn().mockReturnValue('removeQuery'),
      clearQuery: jest.fn().mockReturnValue('clearQuery')
    }
    let wrapper = shallow(
      <Provider store={store}>
        <TransactionTable.WrappedComponent {...props} />
      </Provider>
    ).dive()
    expect(wrapper.instance().props.executeQuery()).toBe('executeQuery');
    expect(wrapper.instance().props.updateQuery()).toBe('updateQuery');
    expect(wrapper.instance().props.removeQuery()).toBe('removeQuery');
    expect(wrapper.instance().props.clearQuery()).toBe('clearQuery');
  });

  test('component did mount', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn()
    }
    let wrapper = mount(
      <Provider store={store}>
        <TransactionTable.WrappedComponent {...props} />
      </Provider>
    )
    expect(props.fetchTx).toHaveBeenCalled();
  });

  test('mapStateToProps with default state', () => {
    const state = {
      queryEngine: {
        query: { last: 15 },
        queryResults: transactions
      },
      chains: {
        selectedChain: 'airline cartel 1'
      }
    }
    expect(mapStateToProps(state)).toMatchSnapshot();
  });

  test('redux form', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      handleSubmit: jest.fn(),
    }

    let wrapper = mount(
      <Provider store={store}>
        <TransactionTable.WrappedComponent {...props} />
      </Provider>, { lifecycleExperimental: true }
    )
    expect(store.getState().form['transaction-query'].values).toBeUndefined()
    const form = wrapper.find('form')
    const select = wrapper.find('Field').first()
    const input = wrapper.find('Field').last()
    expect(input.instance().value).toBe(undefined)
    expect(select.instance().value).toBe(undefined)
    input.simulate('change', { target: { value: '15' } })
    input.simulate('keypress', { key: 'Enter' })
    select.simulate('change', { target: { value: 'BlockNumber' } })
    expect(input.instance().value).toBe('15')
    expect(select.instance().value).toBe('BlockNumber')
    form.simulate('submit')
    expect(store.getState().form['transaction-query'].values).toEqual({ value: '15', query: 'BlockNumber' })

  });

  test('update query method', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      handleSubmit: jest.fn(),
      store: store
    }

    let wrapper = shallow(
      <TransactionTable.WrappedComponent {...props} />
    ).dive().dive().dive()
    wrapper.instance().updateQuery({ value: '15', query: 'BlockNumber' })
    expect(props.updateQuery).toHaveBeenCalledWith('BlockNumber', '15')
  });

  test('update query method without any values', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      handleSubmit: jest.fn(),
      store: store
    }

    let wrapper = shallow(
      <TransactionTable.WrappedComponent {...props} />
    ).dive().dive().dive()
    wrapper.instance().updateQuery({ value: null, query: null })
    expect(props.updateQuery).not.toHaveBeenCalledWith('BlockNumber', '15')
  });

  test('componentWillReceiveProps', () => {
    const props = {
      query: { last: 15 },
      queryResults: [],
      fetchTx: jest.fn(),
      executeQuery: jest.fn(),
      updateQuery: jest.fn(),
      removeQuery: jest.fn(),
      clearQuery: jest.fn(),
      handleSubmit: jest.fn(),
      store: store
    }
    let wrapper = shallow(
      <TransactionTable.WrappedComponent {...props} />
    ).dive().dive().dive()
    const props2 = {
      query: { last: 10 },
      executeQuery: jest.fn()
    }
    wrapper.instance().componentWillReceiveProps(props2)
    expect(props2.executeQuery).toHaveBeenCalled()
  });

})