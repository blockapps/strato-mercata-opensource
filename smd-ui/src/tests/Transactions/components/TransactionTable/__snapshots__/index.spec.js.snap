// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionTable: index mapStateToProps with default state 1`] = `
Object {
  "query": Object {
    "last": 15,
  },
  "queryResults": undefined,
  "selectedChain": "airline cartel 1",
}
`;
